File: server/assets/solutions/ohtu2022/part7/java/KiviPaperiSakset/src/main/java/ohtu/kivipaperisakset/KiviPaperiSakset.java

package ohtu.kivipaperisakset;

import ohtu.kivipaperisakset.tekoaly.NormaaliTekoaly;
import ohtu.kivipaperisakset.tekoaly.TekoalyParannettu;
import java.util.Scanner;

// abstrakti luokka joka sisältää pelin rungon
public abstract class KiviPaperiSakset {
    protected static final Scanner scanner = new Scanner(System.in);
    
    // peliolion luomisen hoitava tehdasmetodi
    public static KiviPaperiSakset peli(String valinta) {
        if ("a".equals(valinta)) {
            return new KPSPelaajaVsPelaaja();
        } else if  ("b".equals(valinta)) {
            return new KPSTekoaly(new NormaaliTekoaly());
        } else{
            return new KPSTekoaly(new TekoalyParannettu(20));
        }  
    }
    
    // itse pelaamisen hoitava template metodi
    public void pelaa() {
        Tuomari tuomari = new Tuomari();

        String ekanSiirto = ensimmaisenSiirto();
        String tokanSiirto = toisenSiirto();
        
        while (onkoOkSiirto(ekanSiirto) && onkoOkSiirto(tokanSiirto)) {
            tuomari.kirjaaSiirto(ekanSiirto, tokanSiirto);
            System.out.println(tuomari);
            System.out.println();
            
            ekanSiirto = ensimmaisenSiirto();
            tokanSiirto = toisenSiirto();
            muistaSiirto(ekanSiirto);
        }

        System.out.println();
        System.out.println("Kiitos!");
        System.out.println(tuomari);
    }
    
    protected String ensimmaisenSiirto() {
        System.out.println("Ensimmäisen pelaajan siirto: ");
        return scanner.nextLine();
    }

    // tämä on abstrakti metodi sillä sen toteutus vaihtelee eri pelityypeissä
    abstract protected String toisenSiirto();
    
    protected static boolean onkoOkSiirto(String siirto) {
        return "k".equals(siirto) || "p".equals(siirto) || "s".equals(siirto);
    }

    // oletusaarvoisesti riittää että ei tehdä mitään
    protected void muistaSiirto(String ekanSiirto) {
    }
}

File: server/assets/solutions/ohtu2023/part7/kivi-paperi-sakset/src/kivi_paperi_sakset.py
from tuomari import Tuomari

class KiviPaperiSakset:
    def pelaa(self):
        tuomari = Tuomari()

        while True:
            ekan_siirto = self._ensimmaisen_siirto()
            tokan_siirto = self._toisen_siirto()    
            
            if not self._onko_ok_siirto(ekan_siirto) or not self._onko_ok_siirto(tokan_siirto):
              break

            tuomari.kirjaa_siirto(ekan_siirto, tokan_siirto)
            print(tuomari)

        print("Kiitos!")
        print(tuomari)

    def _onko_ok_siirto(self, siirto):
        return siirto == "k" or siirto == "p" or siirto == "s"
    
    def _ensimmaisen_siirto(self):
        siirto = input("Ensimmäisen pelaajan siirto: ")
        self._kasittele_ensimmaisen_siirto(siirto)
        return siirto

    def _toisen_siirto(self):
        return input("Toisen pelaajan siirto: ")
    
    def _kasittele_ensimmaisen_siirto(self, siirto):
        pass